pkgs/applications/networking/wgnord/default.nix

```diff
@@ -0,0 +1,54 @@
+{ coreutils
+, curl
+, fetchFromGitHub
+, gnugrep
+, gnused
+, iproute2
+, jq
+, lib
+, makeWrapper
+, stdenv
+, wireguard-tools
+}:
+
+stdenv.mkDerivation rec {
+  pname = "wgnord";
+  version = "0.1.10";
+
+  src = fetchFromGitHub {
+    owner = "phirecc";
+    repo = pname;
+    rev = version;
+    hash = "sha256-T7dAEgi4tGvrzBABGLzKHhpCx0bxSCtTVI5iJJqJGlE=";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  buildPhase = ''
+    substituteInPlace ${pname} \
+      --replace '$conf_dir/countries.txt' "$out/share/countries.txt" \
+      --replace '$conf_dir/countries_iso31662.txt' "$out/share/countries_iso31662.txt"
+  '';
+
+  installPhase = ''
+    install -Dm755 ${pname} -t $out/bin/
+    install -Dm 644 countries.txt -t $out/share/
+    install -Dm 644 countries_iso31662.txt -t $out/share/
+    wrapProgram $out/bin/${pname} --prefix PATH : ${lib.makeBinPath [
+      coreutils
```

Comment:
We don't really need this on nixos. You can only create a system without coreutils by using lib.mkForce on environment.systemPackages and then most things will break https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/system-path.nix#L15

Reply:
sure, but since we are targeting more than just nixos, it is helpful to pin away from system coreutils? this always bites me on darwin.

Reply:
I would suggest to globally install updates of all the gnu tools instead of hardcoding them into way to many packages.

Reply:
if this is a hard requirement, I am happy to apply it, but this has already caused unpredictable behaviour on darwin and bsd systems for me

Reply:
It is unreasonable amounts of work to add those packages to all packages that require them. The simpler and better to maintain solution is to add those packages on systems that do not ship them.

> I would also ask what users on those oses who need the builtin versions for system scripting, and gnu versions for nixpkgs do?

PATH is only expanded in the user session. It would be even more confusing to have different core programs in different programs.

---

Short: I am against hardcoding such core programs into packages, if other distros require them, add them to the nixos installation.

Reply:
> It is unreasonable amounts of work to add those packages to all packages that require them.

If this is so much work, why do we do it this way for every shared object?

> The simpler and better to maintain solution is to add those packages on systems that do not ship them.

I agree it is simpler for nixpkgs, but it leads to heisen bugs and "works on my machine" issuues. I hit this exact issue with find and getopt recently, and it was incredibly confusing to the user. Furthermore, it undermined the "nix workes the same everywhere" promise they had been sold.

> PATH is only expanded in the user session. It would be even more confusing to have different core programs in different programs.

This is exactly requirement for how my corporate macbook works. There are system daemons and scripts that the company maintains, that use the built-in versions of find and other system tools. Then I have nix which runs things that expect both proprietary system utilities (launched) and nix dependencies (gnu find, not bsd find). I would expect being explicit is the only way to make this clear?

> Short: I am against hardcoding such core programs into packages, if other distros require them, add them to the nixos installation.

Can you elaborate on what you mean? Nixos is the one that works ootb, so I am unclear what "add[ing] them to nixos installation" does to resolve the issue, or do you mean the "nix install" should setup PATH to contain everything in stdenv?

Reply:
> should we be pushing maintainers to patch relative and bare paths to absolute paths into all binary executions?

This is, at least for Bash/POSIX Shell, what https://github.com/abathur/resholve exists to do.

Your call, as maintainer/packager, but I made a quick example atop your PR:

https://github.com/abathur/nixpkgs/commit/d6030d4be4a6c8fb084f617d9991b5211405596e

<details><summary>And for reference, the ~resholved script:</summary>

```console
$ cat /nix/store/h76vnv7h9lv3n0v10qfiicx81yb4n3vv-wgnord-0.1.10/bin/wgnord 
Fri Jan 06 2023 20:38:57 --> 
#!/nix/store/4xw8n979xpivdc46a9ndcvyhwgif00hz-bash-5.1-p16/bin/sh
alias query="/nix/store/m2h1p50yvcq5j9b3hkrwqnmrr9pbkzpz-curl-7.86.0-bin/bin/curl -s -H 'User-Agent: NordApp Linux 3.12.0 Linux 5.4.0-58-generic'"
alias print="printf \"\033[32m%s\n\033[0m\""
alias print_error="printf \"\033[31m%s\n\033[0m\""
host="api.nordvpn.com"
conf_dir=/var/lib/wgnord
check_root() {
	if [ "$(/nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/id -u)" -ne 0 ]; then
		echo "This script must be run as root"
		exit 1
	fi
}
login() {
	[ -z "$1" ] && help
	pass="$2"
	if [ -z "$pass" ]; then
		printf "Enter password: "
		/nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/stty -echo
		read pass
		/nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/stty echo
		echo
	fi
	print "Logging in..."
	umask 027
	/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -njc "{username: \"$1\", password: \"$(echo "$pass" | /nix/store/5dv5cq1lwvsijr9p16p2kp79g1dbajk3-gnused-4.8/bin/sed 's/"/\\"/g')\"}" > $conf_dir/account.json
	get_token
	get_credentials
}
get_token() {
	print "Getting new access token..."
	query "https://$host/v1/users/tokens" -d "@$conf_dir/account.json" -H "Content-Type: application/json" > $conf_dir/token.json
	err="$(/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -er '.errors.message' $conf_dir/token.json)" && { print_error "Error getting token: $err"; exit 1; }
}
get_credentials() {
	print "Getting server credentials..."
	token_date=$(/nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/date -d "$(/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -r '.expires_at' $conf_dir/token.json)" --utc +%s)
	current_date=$(/nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/date --utc +%s)
	[ "$current_date" -ge $((token_date-60)) ] && get_token
	auth_token="$(printf "%s" "token:$(/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.token' $conf_dir/token.json)" | /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/base64 -w 0)"
	query "https://$host/v1/users/services/credentials" -H "Content-Type: application/json" -H "Authorization: Basic $auth_token" > $conf_dir/credentials.json
	err="$(/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -er '.errors.message' $conf_dir/credentials.json)" && { print_error "Error getting credentials: $err"; exit 1; }
}
get_country_code() {
    buf="$(/nix/store/bcvccw6y9bfil6xrl5j7psza7hnd16ry-gnugrep-3.7/bin/grep -i "^$1	" -m 1 /nix/store/8pc71rrw00qksrz9l6x313i4xjlw0zp4-wgnord-unresholved-0.1.10/share/countries_iso31662.txt | /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/cut -d "	" -f 2)"
    [ -z "$buf" ] && buf="$(/nix/store/bcvccw6y9bfil6xrl5j7psza7hnd16ry-gnugrep-3.7/bin/grep -i "$1" -m 1 /nix/store/8pc71rrw00qksrz9l6x313i4xjlw0zp4-wgnord-unresholved-0.1.10/share/countries.txt | /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/cut -d "	" -f 2)"
    echo "$buf"
}
connect() {
	out_file=/etc/wireguard/wgnord.conf
	OPTIND=2
	while getopts "fno:" o; do case "$o" in
		f) force=1 ;;
		n) dont_act=1 ;;
		o) out_file="$OPTARG" ;;
		*) exit 1 ;;
	esac done
	[ -z "$1" ] && help
	print "Finding best server..."
	if [ $force ] || [ ! -f $conf_dir/coords.json ] && [ ! $dont_act ]; then
		is_connected && disconnect
		insights="$(query "https://$host/v1/helpers/ips/insights")"
		longitude="$(echo "$insights" | /nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.longitude')"
		latitude="$(echo "$insights" | /nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.latitude')"
		/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -njc "{longitude: $longitude, latitude: $latitude}" > $conf_dir/coords.json
	else
		longitude="$(/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -er '.longitude' $conf_dir/coords.json)"
		latitude="$(/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -er '.latitude' $conf_dir/coords.json)"
	fi
	country_code="$(get_country_code "$1")"
	recommendations="$(query "https://$host/v1/servers/recommendations?limit=20&filters%5Bservers.status%5D=online&filters%5Bservers_technologies%5D=35&filters%5Bservers_technologies%5D%5Bpivot%5D%5Bstatus%5D=online&fields%5Bservers.id%5D&fields%5Bservers.name%5D&fields%5Bservers.hostname%5D&fields%5Bservers.station%5D&fields%5Bservers.status%5D&fields%5Bservers.load%5D&fields%5Bservers.created_at%5D&fields%5Bservers.groups.id%5D&fields%5Bservers.groups.title%5D&fields%5Bservers.technologies.id%5D&fields%5Bservers.technologies.metadata%5D&fields%5Bservers.technologies.pivot.status%5D&fields%5Bservers.specifications.identifier%5D&fields%5Bservers.specifications.values.value%5D&fields%5Bservers.locations.country.name%5D&fields%5Bservers.locations.country.code%5D&fields%5Bservers.locations.country.city.name%5D&fields%5Bservers.locations.country.city.latitude%5D&fields%5Bservers.locations.country.city.longitude%5D&coordinates%5Blongitude%5D=$longitude&coordinates%5Blatitude%5D=$latitude&fields%5Bservers.ips%5D&filters%5Bcountry_id%5D=$country_code")"
	server_name="$(echo "$recommendations" | /nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.[0] | .name')"
	server_hostname="$(echo "$recommendations" | /nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.[0] | .hostname')"
	server_ip="$(echo "$recommendations" | /nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.[0] | .ips[0].ip.ip')"
	server_pubkey="$(echo "$recommendations" | /nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.[0] | .technologies[] | select(.metadata[0].name == "public_key") | .metadata[0].value')"
	privkey="$(/nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq -j '.nordlynx_private_key' $conf_dir/credentials.json)"
	/nix/store/5dv5cq1lwvsijr9p16p2kp79g1dbajk3-gnused-4.8/bin/sed -e "s|PRIVKEY|$privkey|" -e "s|SERVER_PUBKEY|$server_pubkey|" -e "s|SERVER_IP|$server_ip|" $conf_dir/template.conf > "$out_file"
	if [ ! $dont_act ]; then
		print "Connecting to $server_hostname ($server_name)..."
		if is_connected; then
			/nix/store/asdp7lnwclv2s08ajm0v6srmwc1wpv09-wireguard-tools-1.0.20210914/bin/wg-quick strip wgnord | /nix/store/asdp7lnwclv2s08ajm0v6srmwc1wpv09-wireguard-tools-1.0.20210914/bin/wg setconf wgnord /dev/stdin
			/nix/store/asdp7lnwclv2s08ajm0v6srmwc1wpv09-wireguard-tools-1.0.20210914/bin/wg set wgnord fwmark 0xca6c
		else
			/nix/store/asdp7lnwclv2s08ajm0v6srmwc1wpv09-wireguard-tools-1.0.20210914/bin/wg-quick up wgnord
		fi && print "Connected successfully!"
	fi
}
disconnect() {
	print "Disconnecting..."
	/nix/store/asdp7lnwclv2s08ajm0v6srmwc1wpv09-wireguard-tools-1.0.20210914/bin/wg-quick down wgnord
}
is_connected() {
	/nix/store/vs4303zaqb1grq6kgf2z7wsjzlrz71xy-iproute2-6.0.0/bin/ip link show wgnord > /dev/null 2>&1
	return
}
help() { /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/cat << EOF
Usage: wgnord [ l(ogin) | c(onnect) | d(isconnect) | g(et_credentials) ]

login:
    wgnord l "john.smith@example.com" "verysecurepassword"
    If the password was omitted it will be read from stdin
connect:
    wgnord c france
    -f            Refresh cached longitude/latitude
    -n            Don't connect
    -o out.conf   Write config to different file
disconnect:
    wgnord d
get_credentials:
    wgnord g
    This typically isn't needed. You can try running this if the connection fails

wgnord's files are in $conf_dir, edit template.conf to change Wireguard options
EOF
	exit
}

case $1 in
	l|login) check_root; login "$2" "$3" ;;
	c|connect) check_root; shift; connect "$@" ;;
	d|disconnect) check_root; disconnect ;;
	g|get_credentials) check_root; get_credentials ;;
	*) help ;;
esac

### resholve directives (auto-generated) ## format_version: 3
# resholve: fix aliases
# resholve: keep /nix/store/5dv5cq1lwvsijr9p16p2kp79g1dbajk3-gnused-4.8/bin/sed
# resholve: keep /nix/store/asdp7lnwclv2s08ajm0v6srmwc1wpv09-wireguard-tools-1.0.20210914/bin/wg
# resholve: keep /nix/store/asdp7lnwclv2s08ajm0v6srmwc1wpv09-wireguard-tools-1.0.20210914/bin/wg-quick
# resholve: keep /nix/store/bcvccw6y9bfil6xrl5j7psza7hnd16ry-gnugrep-3.7/bin/grep
# resholve: keep /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/base64
# resholve: keep /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/cat
# resholve: keep /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/cut
# resholve: keep /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/date
# resholve: keep /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/id
# resholve: keep /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/stty
# resholve: keep /nix/store/m2h1p50yvcq5j9b3hkrwqnmrr9pbkzpz-curl-7.86.0-bin/bin/curl
# resholve: keep /nix/store/vs4303zaqb1grq6kgf2z7wsjzlrz71xy-iproute2-6.0.0/bin/ip
# resholve: keep /nix/store/vw9lf2d0cfw92dsascyv7q1l5w260r2g-jq-1.6-bin/bin/jq
```

</details>

Aside: PATH wrapping is still a fine approach and is even essential for some scripts. resholve does what it can, but Shell is incredibly dynamic/complex.

Reply:
no, I am a big fan of this solution, as it has the simplixity of `--prefix PATH`, with the effect of `patchelf

Reply:
> Can you elaborate on what you mean? Nixos is the one that works ootb, so I am unclear what "add[ing] them to nixos installation" does to resolve the issue, or do you mean the "nix install" should setup PATH to contain everything in stdenv?

It should be expected that any those packages exist on any system and they are the nix version.
Also soon as programs execute other programs like terminals or shells it is impossible to get this right because then the nix version is always preferred. 

----

Wrapping every package in a manual way is definitely not the solution to this. This must be solved globally otherwise we get stuck in a micro managing rabbit hole.

Reply:
So only curl, iproute2, jq and wireguard-tools should be in this wrapper.

Reply:
we are no longer wrapping anything; please advise